notebooks/e2e/e2e.py

```diff
@@ -0,0 +1,921 @@
+        '&copy; <a href="https://www.openstreetmap.org/copyright">OpenStre'
+        'etMap</a> contributors &copy; <a href="https://carto.com/attribut'
+        'ions">CARTO</a>'
+    )
+
+    # add Carto Positron tile layer
+    folium.TileLayer(
+        name="Carto Positron Basemap",
+        tiles=tiles,
+        attr=attr,
+        show=False,
+        control=True,
+    ).add_to(m)
+
+    # add OpenStreetMap tile layer
+    folium.TileLayer(
+        name="OpenStreetMap Basemap",
+        show=False,
+        control=True,
+    ).add_to(m)
+
+    # handle legend configuration
+    legend_kwds = {}
+    if caption is not None:
+        legend_kwds["caption"] = caption
+    legend_kwds["max_labels"] = max_labels
+
+    # handle setting column layer name in control menu
+    if column_control_name is None:
+        column_control_name = column
+
+    # add data to the map
+    m = gdf.explore(
+        column,
+        m=m,
+        color=color,
+        cmap=cmap,
+        legend_kwds=legend_kwds,
+        name=column_control_name,
+    )
+
+    # add the urban centre layer, if one is provided
+    if uc_gdf is not None:
+        m = uc_gdf.explore(
+            m=m,
+            color="red",
+            style_kwds={"fill": None},
+            name="Urban Centre",
+            show=show_uc_gdf,
+        )
+
+    # add a point marker to the map, if one is provided
+    if point is not None:
+        marker_kwds = {
+            "icon": Icon(
+                color="red",
+                prefix="fa",
+                icon="flag-checkered",
+            )
+        }
+        m = point.explore(
+            m=m,
+            name=point_control_name,
+            marker_type="marker",
+            marker_kwds=marker_kwds,
+            show=show_point,
+        )
+
+        # add in a dashed buffer around the point, if requested
+        if point_buffer is not None:
+            m = (
+                point.to_crs("EPSG:27700")
+                .buffer(point_buffer)
+                .explore(
+                    m=m,
+                    color=point_color,
+                    style_kwds={"fill": None, "dashArray": 5},
+                    name="Max Distance from Destination",
+                    show=show_point,
+                )
+            )
+
+    # add in an extra overlay layer, if requested
+    if overlay is not None:
+        m = overlay.explore(
+            m=m,
+            color="#F46A25",
+            name=overlay_control_name,
+        )
+
+    # get and fit the bounds to the added map layers
+    m.fit_bounds(m.get_bounds())
+
+    # add a layer control button
+    folium.LayerControl().add_to(m)
+
+    # write to file if requested
+    if save is not None:
+        dir_name = os.path.dirname(save)
+        if not os.path.exists(dir_name):
+            os.makedirs(dir_name)
+        m.save(save)
+
+    return m
+
+
+# %%
+# visualise for an ID within the UC
+UC_ID = 4110
+assert UC_ID in travel_times.to_id.unique()
```

Comment:
Use of assert outside of unit tests to be discouraged. See [bandit docs](https://bandit.readthedocs.io/en/latest/plugins/b101_assert_used.html) for more on that. OK with it being in a notebook but wouldn't want to see this in src.